### PR TITLE
More memory and performance optimizations

### PR DESCRIPTION
**Pull request by aslesarenko**
Contains further optimizations motivated by profiling node syncronization (block verification)
Ergo PR: https://github.com/ergoplatform/ergo/pull/1215

### CODE REVIEW

**Review by kushti — CHANGES_REQUESTED** [review]

File: core/src/main/scala/scalan/primitives/Functions.scala [hunk 1]
```diff
@@ -120,7 +120,7 @@ trait Functions extends Base with ProgramGraphs { self: Scalan =>
 
     override lazy val  scheduleIds: DBuffer[Int] = {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply core/src/main/scala/scalan/primitives/Functions.scala:121]: this is override, the method in the base class has ScalaDoc

File: core/src/main/scala/scalan/primitives/NumericOps.scala [hunk 2]
```diff
@@ -23,35 +27,75 @@ trait NumericOps extends Base { self: Scalan =>
...
+  }
 
-  case class IntegralDivide[T](i: ExactIntegral[T])(implicit elem: Elem[T]) extends DivOp[T]("/", i.quot, i)
+  /** Descriptor of binary `/` operation (integral division). */
+  case class IntegralDivide[T](i: ExactIntegral[T])(implicit elem: Elem[T]) extends DivOp[T]("/", i) {
+    override def applySeq(x: T, y: T): T = i.quot(x, y)
+  }
 
-  case class IntegralMod[T](i: ExactIntegral[T])(implicit elem: Elem[T]) extends DivOp[T]("%", i.rem, i)
+  /** Descriptor of binary `%` operation (reminder of integral division). */
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply core/src/main/scala/scalan/primitives/NumericOps.scala:91]: Added ScalaDoc here and also for BigIntIsIntegral

File: core/src/main/scala/scalan/primitives/Thunks.scala [hunk 3]
```diff
@@ -10,6 +10,7 @@ import scalan.util.{Covariant, GraphUtil}
 
 import scala.collection.Seq
 
+/** Slice in Scala cake with definitions of Thunk operations. */
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply core/src/main/scala/scalan/primitives/Thunks.scala:13]: updated ScalaDoc with a  reference to wikipedia

File: core/src/main/scala/scalan/primitives/Thunks.scala [hunk 4]
```diff
@@ -73,7 +74,7 @@ trait Thunks extends Functions with GraphVizExport { self: Scalan =>
     cachedElemByClass(eItem)(classOf[ThunkElem[T]])
   implicit def extendThunkElement[T](elem: Elem[Thunk[T]]): ThunkElem[T] = elem.asInstanceOf[ThunkElem[T]]
 
-  class ThunkDef[A](val root: Ref[A], _scheduleIds: =>ScheduleIds)
+  class ThunkDef[A](val root: Ref[A], _scheduleIds: => ScheduleIds)
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply core/src/main/scala/scalan/primitives/Thunks.scala:77]: documented all public classes and methods in this file.

File: sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala [hunk 5]
```diff
@@ -118,17 +118,16 @@ class ErgoLikeContext(val lastBlockUtxoRoot: AvlTreeData,
       dataBoxes, boxesToSpend, newSpendingTransaction, selfIndex, extension, validationSettings, costLimit, initCost)
 
 
-  override def toSigmaContext(IR: Evaluation, isCost: Boolean, extensions: Map[Byte, AnyValue] = Map()): sigma.Context = {
-    implicit val IRForBox: Evaluation = IR
+  override def toSigmaContext(isCost: Boolean, extensions: Map[Byte, AnyValue] = Map()): sigma.Context = {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala:121]: added more details

**Review by kushti — APPROVED** [review]

**Pull request merged**